File: ListaTarefasFrontEnd/flutter/lista_tarefas/lib/entity/tarefas.dart
import 'dart:js_interop';
import 'package:lista_tarefas/services/model_base.dart';

class Tarefa extends ModelBase {
  int? id;
  String? titulo;
  String? descricao;
  DateTime? datavencimento;
  bool completa;

  Tarefa(
      {required this.id,
      required this.titulo,
      required this.descricao,
      required this.datavencimento,
      required this.completa});

  Map<String, dynamic> toJson() {
    return {
      "id": id ?? 0,
      "titulo": titulo ?? '',
      "descricao": descricao ?? '',
      "dataVencimento": datavencimento.isNull ? null : datavencimento!.toIso8601String(),
      "completa": completa,
    };
  }

  factory Tarefa.fromMap(Map<String, dynamic> map) {
    return Tarefa(
      id: map['id'],
      titulo: map['titulo'],
      descricao: map['descricao'],
      datavencimento: map['datavencimento'],
      completa: map['completa'],
    );
  }

  factory Tarefa.getInstance() {
    return Tarefa(
        id: 0,
        titulo: '',
        descricao: '',
        datavencimento: null,
        completa: false);
  }
  
  @override
  ModelBase fromJson(Map<String, dynamic> json) {
    return Tarefa(
      id: json["id"],
      titulo: json["titulo"],
      descricao: json["descricao"],
      datavencimento: json['dataVencimento'] == null
          ? null
          : DateTime.parse(json['dataVencimento']),
      completa: json["completa"],
    );
  }

  @override
  List<Tarefa> fromJsonToList(Iterable list) {
    return List<Tarefa>.from(list.map((model) => fromJson(model)));
  }
}

File: ListaTarefasFrontEnd/angular/lista-tarefas/src/app/models/tarefas.ts
import { Util } from "../utils/utils";

export class Tarefas {
  constructor(id: number = 0, titulo: string = "", descricao: string = "",
    dataVencimento: string = "", completa: boolean = false)
  {
    this.id = Util.TryValue(id, "number");
    this.titulo = Util.TryValue(titulo, "string");
    this.descricao = Util.TryValue(descricao, "string");
    this.dataVencimento = Util.TryValue(dataVencimento, "string");
    this.completa = Util.TryValue(completa, "boolean");
  }

  id: number;
  titulo: string;
  descricao: string;
  dataVencimento: string;
  completa: boolean;
}
